aiida/engine/processes/ports.py

```diff
@@ -46,28 +40,66 @@ def __init__(self, *args, **kwargs) -> None:
 
     @property
     def non_db_explicitly_set(self) -> bool:
-        """Return whether the a value for `non_db` was explicitly passed in the construction of the `Port`.
+        """Return whether the ``non_db`` keyword was explicitly passed in the construction of the ``InputPort``.
 
-        :return: boolean, True if `non_db` was explicitly defined during construction, False otherwise
+        :return: ``True`` if ``non_db`` was explicitly defined during construction, ``False`` otherwise
         """
         return self._non_db_explicitly_set
 
     @property
     def non_db(self) -> bool:
-        """Return whether the value of this `Port` should be stored as a `Node` in the database.
+        """Return whether the value of this ``InputPort`` should be stored in the database.
 
-        :return: boolean, True if it should be storable as a `Node`, False otherwise
+        :return: ``True`` if it should be stored, ``False`` otherwise
         """
         return self._non_db
 
     @non_db.setter
     def non_db(self, non_db: bool) -> None:
-        """Set whether the value of this `Port` should be stored as a `Node` in the database.
-        """
+        """Set whether the value of this ``InputPort`` should be stored as a ``Data`` in the database."""
         self._non_db_explicitly_set = True
         self._non_db = non_db
 
 
+class WithMetadata:
+    """A mixin that allows an input port to be marked as ``metadata``.
+
+    A ``metadata`` input differs from a normal input as in that it is not linked to the ``ProcessNode`` as a ``Data``
+    node but rather is stored on the ``ProcessNode`` itself (as an attribute, for example).
+    """
+
+    def __init__(self, *args, **kwargs) -> None:
+        self._explicitly_set: bool = bool('metadata' in kwargs)
+        metadata = kwargs.pop('metadata', False)
+        super().__init__(*args, **kwargs)
+        self._metadata: bool = metadata
+
+    @property
+    def metadata_explicitly_set(self) -> bool:
+        """Return whether the ``metadata`` keyword was explicitly passed in the construction of the ``InputPort``.
+
+        :return: ``True`` if ``metadata`` was explicitly defined during construction, ``False`` otherwise
+        """
+        return self._explicitly_set
+
+    @property
+    def metadata(self) -> bool:
```

Comment:
The name was discussed, and the preference was for `metadata` over `is_metadata`, correct?

Reply:
We (@chrisjsewell and @giovannipizzi ) agreed on `metadata` but I have to say we didn't consider a lot of other options. `is_metadata` wasn't brought up for example. Happy to think about the naming if others think this is also a better option.

Reply:
Brought it up during today's meeting and we agree that `is_metadata` is a better name. Have updated the code and commits.